File: DotNetGotchas/CSharp/ES/ESPlatformDiff/ESLib/Factory.cs
// Factory.cs part of ESLib.dll
using System;
using System.EnterpriseServices;

namespace ESLib
{
	[Transaction(TransactionOption.Required), JustInTimeActivation]
	public class Factory : ServicedComponent
	{
		public Comp CreateComp(int key)
		{
			ContextUtil.MyTransactionVote 
				= TransactionVote.Abort;

			Comp theComp = new Comp();
			theComp.init(key);

			ContextUtil.MyTransactionVote 
				= TransactionVote.Commit;

			return theComp;
		}

		protected override void Dispose(bool disposing)
		{
			if (disposing)
			{
				ContextUtil.DeactivateOnReturn = true;
			}
			base.Dispose (disposing);
		}

	}
}

File: DotNetGotchas/VBDN/ES/ESPlatformDiff/ESLib/Factory.vb
' Factory.vb part of ESLib.dll
Imports System.EnterpriseServices

<Transaction(TransactionOption.Required), JustInTimeActivation()> _
Public Class Factory
    Inherits ServicedComponent
    Public Function CreateComp(ByVal key As Integer) As Comp
        ContextUtil.MyTransactionVote = TransactionVote.Abort

        Dim theComp As New Comp

        theComp.init(key)

        ContextUtil.MyTransactionVote = TransactionVote.Commit

        Return theCOmp
    End Function
    Protected Overloads Overrides Sub Dispose( _
        ByVal disposing As Boolean)
        If disposing Then
            ContextUtil.DeactivateOnReturn = True
        End If

        MyBase.Dispose(disposing)
    End Sub
End Class
